### ISSUE TITLE
Proposal: Remove multi-way `if`

### ISSUE THREAD

**Issue by rtfeldman** [elm-lang/elm-plans#3]
## Proposed Changes

**1.** Remove the multi-way `if` syntax.

Currently the Elm Syntax docs include [this example](http://elm-lang.org/docs/syntax#conditionals):

``` elm
if | key == 40 -> n+1
   | key == 38 -> n-1
   | otherwise -> n
```

This would no longer work. Instead, this example would be rewritten to the following:

``` elm
if key == 40 then
    n + 1
else if key == 38 then
    n - 1
else
    n
```

**2.** Remove [`Basics.otherwise`](http://package.elm-lang.org/packages/elm-lang/core/2.1.0/Basics#otherwise), as it exists as a convenience for multi-way `if`.

This would be a breaking compiler change.
## Motivation

The multi-way `if` syntax has these problems:
1. It [can currently lead to non-exhaustive matches](https://github.com/elm-lang/elm-compiler/issues/385).
2. It adds complexity to the compiler and learning curve to new users, in exchange for conciseness but no additional capabilities.
3. Expanding from a normal `if`/`else` to a multi-way `if` requires rewriting the entire expression, and failing to do so when a third case arises leads to stylistic inconsistencies across the code base.
4. It is a rarely-used feature in a language with pattern matching.

**Comment by mgold** [comment]
I thought I wouldn't like chaining 2-way ifs but now that I see it, it looks really good. We go from a hard-to-remember, little-used syntax, to one that extends so naturally from the 2-way if that it already works. Implementation should be almost entirely removal. +1

**Comment by Apanatshka** [comment]
I was actually thinking in the other direction, remove the two-way one. That opens more syntax options to improve the multi-way one. I guess I'll open a counter-proposal, if I can find the time for it. 
My thoughts on your motivation:
1. That's just a bug that needs fixing. 
2. Conciseness doesn't have to add to the learning curve of new users. Elm syntax is alien for most people anyway, so removing this won't really help there. 
3. Yes, but removing the two-way if will result in the same effect.
4. Debatable. Some code is heavy on the union types and therefore pattern matching. This may be an argument for eliminating one of the options since having two ways to write an `if` is silly, but not for removing the multi-way if in particular. 

What I conclude is that those arguments that hold are not specific to removing the multi-way `if` but more to removing one of the two `if` options.

**Comment by rtfeldman** [comment]
> Conciseness doesn't have to add to the learning curve of new users. Elm syntax is alien for most people anyway, so removing this won't really help there. 

The way I look at it is that Elm syntax being alien is a generally negative thing, and changes that make it less alien are generally positive things unless they sacrifice important language characteristics.

People are used to `if` / `else if` / `else` from a multitude of mainstream industry languages, so there's nothing really alien there. The pipe syntax exists in no mainstream industry languages, which makes it feel alien to many.

There's a cost to that, and whatever benefits it brings don't seem to outweigh the cost.

**Comment by mgold** [comment]
The thing I like about the two-way syntax is that it's impossible to have an non-exhaustive analysis. I don't think a multi-way construct could make that guarantee. Using the chained form, it's now impossible in the multi-way as well. This is cited as a benefit in the proposal, but it does mean you may have some cases where you know the pattern match is exhaustive but the conditions don't line up. I suppose this is code smell, so it's good to force people away from it. One fewer runtime error.

Also, would we drop `otherwise` from Basics?

**Comment by Apanatshka** [comment]
@rtfeldman The pipes is something I was going to suggest removing once the two-way `if` is eliminated, since the multi-way if wouldn't be ambiguous with it any more then.

@mgold If you change the multi-way if to enforce the otherwise branch, you get basically the same. 
By the way: the case where you see that the last `else` branch can't happen, you can add a `Debug.crash`. Though I'd prefer to have a different function, perhaps a value `unreachable : a` in `Basics` that prints file, line and column when it gets hit. That's better documentation than doing that by hand in the `Debug.crash` string.

**Comment by texastoland** [comment]
Good catch! I assume so. There's several low hanging issues in Core. Anyone can PR if it's overlooked.

**Comment by mgold** [comment]
@Apanatshka I'd like to see that added behavior added to `Debug.crash` but I don't think we should have  a similar function in Basics. It seems contrary to the non-runtime-errors thing, and like I said this is probably code smell anyway.

@dnalot You're referring to removing `otherwise`, right? We can do that PR once the compiler one is ready. It seems the only use of multi-way if in core is in `Color.elm`, so the same PR can fix that.

**Comment by soupi** [comment]
I agree with @Apanatshka on all points.
I think the multi-way if is much easier to read and write, and I do use it even though I don't write too much Elm code. IMO that both guards/multi-way-ifs and pattern matching are important and serves different causes.

**Comment by evancz** [comment]
The reason I initially brought this idea up with @rtfeldman was that I have been trying to get to a uniform indentation style for Elm, where you can always be 2-space or 4-space indent. (Another goal is "always bring things down a line after `->` and `=` so the example in the initial issue is not following the intent of [the style guide](http://elm-lang.org/docs/style-guide)) Anyway, this is all problematic for the multi-way if. Here it is attempting a 2-space:

``` elm
if  | key == 40 ->
      x + 1
    | key == 38 ->
      x - 1
    | otherwise ->
      x
```

Needing the two spaces after the `if` is kind of sucky. And having the condinion and the branch align exactly seems kind of bad too.

After that I'd say @rtfeldman's points 1 and 3 are the ones that matter most to me.

Anyway, just wanted to say: show code examples! If you think one way is better, show a specific example! If no such example exists, we are all set!

**Comment by evancz** [comment]
Also, if folks are interested in guarded patterns in `case`, multi-way if does not address that. I think that can be useful at times, but it seems like a separate issue to me.

**Comment by Apanatshka** [comment]
Ok, I wrote a counter-proposal in #6. It surprised me that the syntax I came up with is very close to normal nested `if`s, so it could be seen as removing multi-way `if` and adding some syntactic sugar...

**Comment by ryepesg** [comment]
I was thinking of using indentation to avoid `then`:

```
if a <= 0
    Nothing
else if a == 1
    Just x
else
    nthElement xs (a-1)
```

But then I realised that if-then-else is not the idiomatic way for functional programmers (Haskell, F#, Clojure), see [F# control-flow-expressions](http://fsharpforfunandprofit.com/posts/control-flow-expressions/).

Current Elm's conditionals are really great, because of the similar syntax for `if` and `case of`, and the fact we can drop support for Haskell guards (the preferred way in Haskell, see [guards-vs-if-then-else-vs-cases-in-haskell](http://stackoverflow.com/questions/9345589/guards-vs-if-then-else-vs-cases-in-haskell)) without impact in the elegance.

I think the issue "It can currently lead to non-exhaustive matches" could be addressed with a warning. So for me, the relevant issue is "Expanding from a normal if/else to a multi-way if requires rewriting the entire expression, and failing to do so when a third case arises leads to stylistic inconsistencies across the code base". For that case, the normal if/else could better change to:

`if | key == 40 -> n+1 | else -> n`

The arrow is similar in the use of `case of` so I think it is idiomatic. The pipe is not alien, it is used normally in Haskell guards and in the |> <| operators.

New conditions doesn't require rewriting the entire expression:

```
if | key == 40 -> n+1
   | key == 38 -> n-1
   | else      -> n
```

Look how similar is the next example compared to the haskell's normal guard showed in the last link:

```
if | a <= 0 -> Nothing                                      
   | a == 1 -> Just x
   | else   -> nthElement xs (a-1)  
```

**Comment by LiamGoodacre** [comment]
Not sure how I feel about it, but a possible option could be to make `if` be
a synthetic function of type `forall a. Bool -> Lazy a -> Lazy a -> a`.

With the following example use case:

``` elm
ex0 x =
  if (x > 0)
    "pos" <|
  if (x < 0)
    "neg" <|
  else
    "zero"

ex1 x = if (x == 0) "zero" "non-zero"
```

A working (non-lazy) implementation:

``` elm
iF : Bool -> a -> a -> a
iF b l r = if b then l else r

eLSE : a -> a
eLSE = identity

ex0 x =
  iF (x > 0)
    "pos" <|
  iF (x < 0)
    "neg" <|
  eLSE
    "zero"

ex1 x = iF (x == 0) "zero" "non-zero"
```

**Comment by Apanatshka** [comment]
@ryepesg guards on pattern matches are quite different when you involve more patterns, it's not always as easy to translate between that and Elm's multi-way if. 
We're not optimising the syntax for Haskell users, the common user we expect and try to cater is more from the front-end web world of HTML and JavaScript. 

@LiamGoodacre I like the idea of going back to basics and use functions, because this is usually so powerful. But in this case I'm hesitant because of the syntactic overhead of writing lazy functions, and the cognitive overhead of explaining laziness when all a beginner wants is a simple `if`.

**Comment by evancz** [comment]
I was reminded about another issue. If we allow expressions in record update syntax, multi-way ifs are ambiguous. Perhaps this is bad enough to make it impossible. `{ if | x == 1 -> point | x <- 4 }`

**Comment by rtfeldman** [comment]
Ouch...

**Comment by evancz** [comment]
Yeah, that's actually a part of why Elm does not allow arbitrary expressions in that position. I didn't know how to deal with this case (and others), so I decided to answer the question later.

**Comment by mgold** [comment]
Then I think we should drop multi-way if and see if we can rework record update syntax, possibly as part of removing add and remove, to accept arbitrary expressions.

**Comment by rtfeldman** [comment]
+1 to everything @mgold said!

**Comment by bbugh** [comment]
I'm a web technology adult educator with an interest in introducing Elm into upcoming curriculum (because it's amazing and solves a big problem on the web). As I've been learning Elm myself, I keep thinking "how would I teach this?" The most common reason why students falter and give up is not because of one confusing moment, it's because of a "death by papercuts" series of small misunderstandings. 

I think the `else if` is the easiest one explain to a new person. If a goal of Elm is to be inclusive of Ruby and JS developers (which is what I've inferred) then I think it's pretty important to err on the side of consistency with those languages over Haskell, especially for something as basic as a branching `if`. 

As a non-Haskell knowing person, Elm seems to have a bit of "operator-itis". It's particularly confusing when operators have different meanings depending on context (like `|`, `->`). I'm a fan of removing the multi-way if syntax and falling back on `else if`, because 1) it's clear what it's doing, and 2) that's one less single-use edge case for `|`.

As for white space importance in an `if` statement, I think the "sometimes white space matters, sometimes it doesn't" part of Elm is also one of the more confusing aspects. Internal consistency is really important for learning—people look for patterns relatable to what they already know when learning something new. Any changes that reduce those inconsistent edge cases through (simple) explicit language constructs is a big win for adoption in my opinion.

Here's what I'm proposing in case that wasn't clear:

``` elm
if key == 40 then
    n + 1
else if key == 38 then
    n - 1
else
    n

-- or what I would expect is more "elm like"
if key == 40
then n + 1
else if key == 38 
then n - 1
else n
```

**Comment by rtfeldman** [comment]
@bbugh Thanks for your perspective! The first style you mentioned (with `then` at the end of the line instead of at the start) is what I've been doing, and I'm definitely a fan.

**Comment by evancz** [comment]
@bbugh, thanks for you perspective! The style you recommend was the one I initially showed to @rtfeldman that initiated this issue. That style is what I was planning to recommend if this feature was removed for pretty much exactly the reasons you outline.

Also, the one you theorized is "more elm like", not in my code I hope, that looks awful to me! :)

**Comment by evancz** [comment]
For reference, I think we'll do a bit of syntax removal in 0.16 for this and for record extension and deletion. That's what I have in my mind at least.

**Comment by mgold** [comment]
As for style, I like the first way as well. I've been known to put `if`, `then`, and `else` on three lines but only for single way if.

You can also take the opposite perspective: if we only had the single way syntax but it worked just fine for multiway, would we feel the need to add syntax for the multi way? I doubt it. And the thing I really like about this proposal is that the single way _already_ works for multiway.

Although I might be beating a dead horse given that Evan seems to have tentatively accepted this for 0.16.

**Comment by Apanatshka** [comment]
Since this is planned for 0.16, perhaps it's time to close this issue and mark it as accepted?

**Comment by evancz** [comment]
Closing down this repo. This is something I am vaguely planning to do. I will bring it up on the mailing list at the appropriate time so we can show everyone the reasoning and such. I would like to frame and guide that discussion when the time comes.
